### PR TITLE
Update index operations

### PR DESCRIPTION
**Pull request by rohanshah18**
## Problem

After adding open api generated code for control plane, there's a need for a wrapper for the index operations.

## Solution

This PR adds a wrapper for index operations and associated integration tests. Couple of the integration tests are currently commented out and will be handled in the upcoming PR's.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Infrastructure change (CI configs, etc)
- [ ] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan

Ran integration tests.

### CODE REVIEW

File: src/integration/java/io/pinecone/helpers/AssertRetry.java [hunk 1]
```diff
@@ -8,6 +8,10 @@ public class AssertRetry {
     private static int delay = 1500;
 
     public static void assertWithRetry(AssertionRunnable assertionRunnable) throws InterruptedException {
+        assertWithRetry(assertionRunnable, 2);
```
Comment (rohanshah18) [inline src/integration/java/io/pinecone/helpers/AssertRetry.java:11]: Added a new assert with retry function without backoff

File: src/integration/java/io/pinecone/helpers/IndexManager.java [hunk 2]
```diff
@@ -12,69 +11,79 @@
 public class IndexManager {
     private static PineconeClientConfig config;
 
-    public static PineconeConnection createIndexIfNotExistsDataPlane(int dimension) throws IOException, InterruptedException {
-        config = new PineconeClientConfig()
-                .withApiKey(System.getenv("PINECONE_API_KEY"))
-                .withEnvironment(System.getenv("PINECONE_ENVIRONMENT"));
+    public static PineconeConnection createIndexIfNotExistsDataPlane(int dimension, String indexType) throws IOException, InterruptedException {
```
Comment (rohanshah18) [inline src/integration/java/io/pinecone/helpers/IndexManager.java:14]: This class required refactoring based on the new index operations.

File: src/integration/java/io/pinecone/helpers/IndexManager.java [hunk 3]
```diff
@@ -12,69 +11,79 @@
...
         int i = 0;
-        while (i < indexList.size()) {
-            IndexMeta indexMeta = isIndexReady(indexList.get(i), controlPlaneClient);
-            if (indexMeta.getDatabase().getDimension() == dimension
-                    && (indexMeta.getDatabase().getPodType().equals("p1.x1") || config.getEnvironment().equals("gcp-starter"))) {
-                return indexList.get(i);
+        List<IndexModel> indexModels = indexList.getIndexes();
+        while (i < indexModels.size()) {
+            IndexModel indexModel = isIndexReady(indexModels.get(i).getName(), controlPlaneClient);
+            // ToDo: add pod type support
```
Comment (rohanshah18) [inline src/integration/java/io/pinecone/helpers/IndexManager.java:45]: In the future PR, I'd like to test for gcp-starter s.t. no new indexes are created if the gcp-starter is detected.

File: src/integration/java/io/pinecone/integration/controlPlane/index/pod/ConfigureIndexTest.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
+//package io.pinecone.integration.controlPlane.index.pod;
```
Comment (rohanshah18) [inline src/integration/java/io/pinecone/integration/controlPlane/index/pod/ConfigureIndexTest.java:1]: Next PR will target this refactoring.

File: src/integration/java/io/pinecone/integration/dataplane/UpdateAndQueryTest.java [hunk 5]
```diff
@@ -56,6 +57,7 @@ public void UpdateRequiredParamsFetchAndQuerySync() throws InterruptedException
                 .build();
 
         blockingStub.update(updateRequest);
+        Thread.sleep(3500);
```
Comment (rohanshah18) [inline src/integration/java/io/pinecone/integration/dataplane/UpdateAndQueryTest.java:60]: When I switch to data plane refactoring, the PR will update thread.sleep with assertWithRetry.

File: src/main/java/io/pinecone/PineconeClientConfig.java [hunk 6]
```diff
@@ -1,14 +1,10 @@
 package io.pinecone;
 
-import io.pinecone.exceptions.PineconeException;
 import io.pinecone.exceptions.PineconeValidationException;
```
Comment (rohanshah18) [inline src/main/java/io/pinecone/PineconeClientConfig.java:3]: Removed unnecessary imports

File: src/main/java/io/pinecone/PineconeIndexOperationClient.java [hunk 7]
```diff
@@ -1,136 +1,88 @@
...
-    private PineconeIndexOperationClient(PineconeClientConfig clientConfig, OkHttpClient client, String url) {
-        this.client = client;
-        this.clientConfig = clientConfig;
-        this.url = url;
-    }
-
-    public PineconeIndexOperationClient(PineconeClientConfig clientConfig, OkHttpClient client) {
-        this(clientConfig, client, createUrl(clientConfig));
-    }
+    private ManageIndexesApi manageIndexesApi;
```
Comment (rohanshah18) [inline src/main/java/io/pinecone/PineconeIndexOperationClient.java:16]: Allow users to pass in config as well

File: src/test/java/io/pinecone/PineconeIndexOperationClientTest.java [hunk 8]
```diff
@@ -1,347 +1,347 @@
...
-        when(mockClient.newCall(any(Request.class))).thenReturn(mockCall);
-        PineconeIndexOperationClient indexOperationClient = new PineconeIndexOperationClient(clientConfig, mockClient);
-        indexOperationClient.configureIndex(indexName, configureIndexRequest);
-
-        verify(mockClient, times(1)).newCall(any(Request.class));
-        verify(mockCall, times(1)).execute();
-    }
-}
\ No newline at end of file
+//package io.pinecone;
```
Comment (rohanshah18) [inline src/test/java/io/pinecone/PineconeIndexOperationClientTest.java:1]: Next PR will handle this.

File: src/integration/java/io/pinecone/integration/controlPlane/index/pod/CreateListAndDeleteIndexTest.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        assertEquals(indexName, indexModel.getName());
+        assertEquals(IndexMetric.COSINE, indexModel.getMetric());
+
+        // List the index
+        IndexList indexList = controlPlaneClient.listIndexes();
+        assert !Objects.requireNonNull(indexList.getIndexes()).isEmpty();
+
+        // Delete the index
+        controlPlaneClient.deleteIndex(indexName);
+        Thread.sleep(3500);
```
Comment (rohanshah18) [inline src/integration/java/io/pinecone/integration/controlPlane/index/pod/CreateListAndDeleteIndexTest.java:51]: Will replace it with assertwithretry() maybe

Comment (rohanshah18) [reply src/integration/java/io/pinecone/integration/controlPlane/index/pod/CreateListAndDeleteIndexTest.java:51]: Also planning to add more tests for each functionality

File: src/integration/java/io/pinecone/integration/controlPlane/index/serverless/CreateDescribeListAndDeleteIndexTest.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        assertEquals(indexName, indexModel.getName());
+        assertEquals(IndexMetric.COSINE, indexModel.getMetric());
+
+        // List the index
+        IndexList indexList = controlPlaneClient.listIndexes();
+        assert !Objects.requireNonNull(indexList.getIndexes()).isEmpty();
+
+        // Delete the index
+        controlPlaneClient.deleteIndex(indexName);
+        Thread.sleep(3500);
```
[the remark this answers is not in the document]
Comment (rohanshah18) [reply src/integration/java/io/pinecone/integration/controlPlane/index/serverless/CreateDescribeListAndDeleteIndexTest.java:52]: Also planning to add more tests for each functionality

**Comment by rohanshah18** [comment]
@austin-denoble Overall, this change will unblock you with adding collections and more refactoring are possible in future as we discussed internally. But any of the classes you're dealing with, are finalized as a part of this PR.

File: src/main/java/io/pinecone/PineconeIndexOperationClient.java [hunk 11]
```diff
@@ -1,136 +1,88 @@
...
-import java.util.List;
+import org.openapitools.client.ApiClient;
+import org.openapitools.client.ApiException;
+import org.openapitools.client.api.ManageIndexesApi;
+import org.openapitools.client.model.ConfigureIndexRequest;
+import org.openapitools.client.model.CreateIndexRequest;
+import org.openapitools.client.model.IndexList;
+import org.openapitools.client.model.IndexModel;
 
 public class PineconeIndexOperationClient {
```
Comment (rohanshah18) [inline src/main/java/io/pinecone/PineconeIndexOperationClient.java:14]: Going to rename this. For now, I'm thinking PineconeControlPlaneClient.

File: src/main/java/io/pinecone/PineconeIndexOperationClient.java (hunk 7 above)
Comment (austin-denoble) [reply src/main/java/io/pinecone/PineconeIndexOperationClient.java:16]: Is this primarily for flexibility while we iterate on the approach? It seems like all it's used for at the moment is passing around the `apiKey`, but I can see we still need it when performing data plane operations.

I think this is fine for now, but my opinion is we should look at refactoring the `PineconeClient` and `PineconeConnectionConfig` when we're looking at the data plane to fully remove `PineconeClientConfig`.

File: src/integration/java/io/pinecone/integration/controlPlane/index/pod/CreateListAndDeleteIndexTest.java (hunk 9 above)
Comment (austin-denoble) [reply src/integration/java/io/pinecone/integration/controlPlane/index/pod/CreateListAndDeleteIndexTest.java:51]: I know you mentioned adding more tests, but we should probably assert that the delete worked here.

If not we can probably get rid of the `sleep()` completely for now as I don't think it's necessary unless you're going to be checking if the index was actually deleted.

File: src/integration/java/io/pinecone/integration/controlPlane/index/serverless/CreateDescribeListAndDeleteIndexTest.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import org.junit.jupiter.api.Test;
+import org.openapitools.client.model.*;
+
+import java.util.Objects;
+
+import static io.pinecone.helpers.IndexManager.isIndexReady;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+public class CreateDescribeListAndDeleteIndexTest {
```
Comment (austin-denoble) [inline src/integration/java/io/pinecone/integration/controlPlane/index/serverless/CreateDescribeListAndDeleteIndexTest.java:15]: **nit:** Can we align the names of the `/pod` and `/serverless` test classes? They diverge a bit and it seems like they're going to be doing very similar things.

It's called `CreateListAndDeleteIndexTest` for `/pod/`.

**Review by austin-denoble — APPROVED** [review]
Nice job, overall this looks good to me as a first step towards refactoring everything around the generated code, I think will unblock us to continue with the integration tests and then data plane refactoring.

Main concerns for me right now are making sure we make the UX reasonable for users. I think there's a bunch we could do to refactor `PineconeClient`, `PineconeClientConfig`, `PineconeConnection`, and `PineconeConnectionConfig` into something more reasonable, because it feels like a lot to keep track of right now.

In the other clients, we have a top-level `Pinecone` class that allows you to perform control plane operations, and then that class abstracts a lot of the internals for hitting the dataplane, so users can more easily target and work with indexes without a bunch of setup. I think we should try and shoot for something similar in Java, does that make sense? We can discuss further offline.

File: src/integration/java/io/pinecone/integration/controlPlane/index/pod/CreateListAndDeleteIndexTest.java (hunk 9 above)
Comment (rohanshah18) [reply src/integration/java/io/pinecone/integration/controlPlane/index/pod/CreateListAndDeleteIndexTest.java:51]: lol that's the reason I started this thread :)

File: src/integration/java/io/pinecone/integration/controlPlane/index/serverless/CreateDescribeListAndDeleteIndexTest.java (hunk 12 above)
Comment (rohanshah18) [reply src/integration/java/io/pinecone/integration/controlPlane/index/serverless/CreateDescribeListAndDeleteIndexTest.java:15]: Yeah its going to be refactored into individual functions if I can.

File: src/main/java/io/pinecone/PineconeIndexOperationClient.java (hunk 7 above)
Comment (rohanshah18) [reply src/main/java/io/pinecone/PineconeIndexOperationClient.java:16]: Yes, I'll be refactoring all of the other classes as a part of future PR's. This is mainly to unblock you for collections work.

**Pull request merged**